packages/ifpack2/test/unit_tests/Ifpack2_UnitTestDatabaseSchwarz.cpp

```diff
@@ -0,0 +1,177 @@
+{
+//we are now in a class method declared by the above macro, and
+//that method has these input arguments:
+//Teuchos::FancyOStream& out, bool& success
+
+  typedef LocalOrdinal LO;
+  typedef GlobalOrdinal GO;
+
+  typedef Tpetra::CrsMatrix<Scalar,LO,GO,Node> crs_matrix_type;
+  typedef Tpetra::Map<LO,GO,Node> map_type;
+  typedef Tpetra::MultiVector<Scalar,LO,GO,Node> MV;
+  typedef Tpetra::RowMatrix<Scalar,LO,GO,Node> row_matrix_type;
+  typedef Teuchos::ScalarTraits<Scalar> STS;
+  const Scalar one = STS::one ();
+  const Scalar two = STS::one () + STS::one ();
+
+  out << "Ifpack2::DatabaseSchwarz: Test0" << endl;
+  Teuchos::OSTab tab1 (out);
+
+  global_size_t num_rows_per_proc = 5;
+
+  out << "Creating row Map and CrsMatrix" << endl;
+
+  RCP<const map_type> rowmap =
+    tif_utest::create_tpetra_map<LO, GO, Node> (num_rows_per_proc);
+  RCP<const crs_matrix_type> crsmatrix =
+    tif_utest::create_test_matrix<Scalar, LO, GO, Node> (rowmap);
+
+  out << "Creating DatabaseSchwarz instance" << endl;
+
+  Ifpack2::DatabaseSchwarz<row_matrix_type> prec (crsmatrix);
+  Teuchos::ParameterList params, zlist;
+
+  out << "Filling in ParameterList for DatabaseSchwarz" << endl;
+
+  params.set ("database schwarz: patch size", static_cast<typename Ifpack2::DatabaseSchwarz<row_matrix_type>::local_ordinal_type>(9));
+  params.set ("database schwarz: patch tolerance", 1e-6);
+  params.set ("database schwarz: skip database", false);
+  params.set ("database schwarz: print database summary", true);
+
+  out << "Setting DatabaseSchwarz's parameters" << endl;
+
+  TEST_NOTHROW(prec.setParameters(params));
+
+  out << "Testing domain and range Maps of DatabaseSchwarz" << endl;
+
+  // FIXME (mfh 26 Jul 2015) The domain and range Maps of the
+  // preconditioner don't have to be the same object; they only need
+  // to be the same in the sense of Tpetra::Map::isSameAs().
+  //
+  //trivial tests to insist that the preconditioner's domain/range maps are
+  //identically those of the matrix:
+  const map_type* mtx_dom_map_ptr = &*crsmatrix->getDomainMap();
+  const map_type* mtx_rng_map_ptr = &*crsmatrix->getRangeMap();
+  const map_type* prec_dom_map_ptr = &*prec.getDomainMap();
+  const map_type* prec_rng_map_ptr = &*prec.getRangeMap();
+  TEST_EQUALITY( prec_dom_map_ptr, mtx_dom_map_ptr );
+  TEST_EQUALITY( prec_rng_map_ptr, mtx_rng_map_ptr );
+
+  out << "Calling DatabaseSchwarz's initialize()" << endl;
+  prec.initialize();
+
+  out << "Calling DatabaseSchwarz's compute()" << endl;
+  prec.compute();
+
+  MV x (rowmap, 2), y (rowmap, 2), z (rowmap, 2);
+  x.putScalar (one);
+
+  out << "Calling DatabaseSchwarz's apply()" << endl;
+  prec.apply(x, y);
+
+  // The solution should now be full of 1/2s
+  z.putScalar (one / two);
+
+  Teuchos::ArrayRCP<const Scalar> yview = y.get1dView();
+  Teuchos::ArrayRCP<const Scalar> zview = z.get1dView();
+
+  //TEST_COMPARE_FLOATING_ARRAYS(yview, zview, 4*STS::eps ());
```

Comment:
Actually test the result?

Reply:
I don't have any good examples of analytic solutions that I can port into this testing framework without using something like a gold file

Reply:
I was thinking one good test is the `"skip database" = true` route should agree with the `"skip database" = false` route

Reply:
I'd just cook up some simple problem with a known solution and a couple of patches.

Reply:
Okay I figured it out. I'm not sure what the default matrix is, but I got an analytic solution, so I'm happy enough